lib/minio_runner/minio_server_manager.rb

```diff
@@ -0,0 +1,101 @@
+# frozen_string_literal: true
+
+require_relative "child_process"
+
+module MinioRunner
+  class MinioServerManager
+    SERVER_STOP_TIMEOUT_SECONDS = 5
+
+    attr_reader :pid, :process
+
+    class << self
+      def start
+        @server = new
+        @server.start
+      end
+
+      def stop
+        return if @server.nil?
+        @server.stop
+      end
+    end
+
+    def start
+      if process_running?
+        MinioRunner.logger.debug("Already started minio server.")
+        return
+      end
+
+      MinioRunner.logger.debug("Starting minio server...")
+
+      MinioRunner::System.exit_hook { stop }
+
+      @process =
+        MinioRunner::ChildProcess.build(
+          server_command,
+          env: {
+            "MINIO_ROOT_USER" => MinioRunner.config.minio_root_user,
+            "MINIO_ROOT_PASSWORD" => MinioRunner.config.minio_root_password,
+            "MINIO_DOMAIN" => MinioRunner.config.minio_domain,
+          },
+          log_file: log_file_path,
+        )
+
+      @process.start
+
+      # Make sure the minio server is ready to accept requests.
+      health_check(retries: 3)
+
+      MinioRunner.logger.debug("minio server running at pid #{@process.pid}!")
+    end
+
+    def stop
+      return if process_exited?
+
+      MinioRunner.logger.debug("Stopping minio server running at pid #{@pid}...")
+      @process.stop(SERVER_STOP_TIMEOUT_SECONDS)
+      @process = nil
+      MinioRunner.logger.debug("minio server stopped")
+    end
+
+    def process_running?
+      defined?(@process) && @process&.alive?
+    end
+
+    def process_exited?
+      @process.nil? || @process.exited?
+    end
+
+    private
+
+    def server_command
+      command = []
+
+      # server start command for minio
+      command << "#{MinioRunner::MinioBinary.binary_file_path} server #{MinioRunner.config.minio_data_directory}"
+
+      # flags for minio
+      command << "--console-address :#{MinioRunner.config.minio_console_port}"
+      command << "--address :#{MinioRunner.config.minio_port}"
```

Comment:
I wonder if we should specify `localhost:#{port}` by default, so that the server does not become available on the machine's external (and potentially untrusted) network interface? And then we could provide an option for people to reconfigure configure the bind address if they want the server available externally.

(mostly thinking of the Discourse rspec use case here - we don't want to be exposing services on external network interfaces of dev machines)

Reply:
I _think_ it doesn't matter in this case because we are specifying `MINIO_DOMAIN`, however I'll double check. Maybe this is not even needed; I think without it minio starts at port 9000 which is fine for us.

Reply:
I just fixed this to use `minio_domain` before the port